src/lib/configs/project/spell.ts

```diff
@@ -0,0 +1,294 @@
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import { access } from "fs/promises";
+import { dirname, resolve } from "path";
+
+import type { JSONSchemaType } from "ajv";
+
+import {
+  FLUENCE_CONFIG_FILE_NAME,
+  SPELL_CONFIG_FILE_NAME,
+  TOP_LEVEL_SCHEMA_ID,
+  U32_MAX,
+} from "../../const.js";
+import { ensureSpellAbsolutePath } from "../../helpers/downloadFile.js";
+import { validateBatch } from "../../helpers/validations.js";
+import { ensureFluenceDir } from "../../paths.js";
+import {
+  getConfigInitFunction,
+  InitConfigOptions,
+  InitializedConfig,
+  InitializedReadonlyConfig,
+  getReadonlyConfigInitFunction,
+  Migrations,
+  ConfigValidateFunction,
+} from "../initConfig.js";
+
+const MAX_PERIOD_YEAR = 100;
+
+/** Max period is 100 years in secs: 60 sec * 60 min * 24 hours * 365 days * 100 years */
+const MAX_PERIOD_SEC = 60 * 60 * 24 * 365 * MAX_PERIOD_YEAR;
+
+type SpellProperties = {
+  aquaFilePath: string;
+  function: string;
+  initArgs?: Record<string, unknown>;
+  clock?: {
+    periodSec?: number;
+    startTimestamp?: string;
+    endTimestamp?: string;
+    startDelaySec?: number;
+    endDelaySec?: number;
+  };
+};
+
+export type OverridableSpellProperties = Partial<SpellProperties>;
+
+export type ConfigV0 = {
+  version: 0;
+} & SpellProperties;
+
+const spellProperties = {
+  version: { type: "number", const: 0 },
+  aquaFilePath: {
+    type: "string",
+    description:
+      "Path to Aqua file which contains an Aqua function that you want to use as a spell",
+  },
+  function: {
+    type: "string",
+    description: "Name of the Aqua function that you want to use as a spell",
+  },
+  initArgs: {
+    type: "object",
+    description:
+      "A map of Aqua function arguments names as keys and arguments values as values. They will be passed to the spell function. Also they will be stored in the key-value storage for this particular spell. It's possible to change argument values inside the spell itself if you return a new map of new arguments from the spell function. Also it's possible to update these values from another spell using Spell.set_string method",
+    nullable: true,
+  },
+  clock: {
+    type: "object",
+    nullable: true,
+    description: `Trigger the spell execution periodically. If you want to disable this property by overriding it in ${FLUENCE_CONFIG_FILE_NAME} - pass empty config for it like this: \`clock: {}\``,
+    properties: {
+      periodSec: {
+        type: "number",
+        description:
+          "How often the spell will be executed. If set to 0, the spell will be executed only once. If this value not provided at all - the spell will never be executed",
```

Comment:
> If this value _is_ not provided at all - the spell will never be executed

Is this part of fcli logic?

Since you cannot _not_ provide period to the rust-peer.

Reply:
Yes it's part of the logic I wrote. The reason is - I don't know any other good way to succinctly express that you wanna override the property in yaml (json form) - this enables passing empty object for clock property in fluence.yaml spell overrides to express that you don't wanna run clock at all (even though spell definition includes it)